### PR TITLE
feat: Or validator

### LINKED ISSUE #145: Exclude approval requirement for certain PRs
It'd be nice if we could remove the requirement for approvals for certain PRs. In particular, I'd be interesting in making PRs created by dependabot to not require approvals. This could be done based on a label, e.g.

```yaml
mergeable:
  pull_requests:
    approvals:
      min: 5
      except:
         label: dependencies
```

If this is something that would be accepted, I can work on a PR to implement it.

**EDIT**:

Actually, I see there's an undocumented `and`/`or` operator that could be used to implement this. Then all that'd be needed is adding a 'label' rule to processFilters.

```yaml
mergeable:
  pull_requests:
    approvals:
      or:
        - min: 5
        - label: '^dependencies$'
```

### PR DESCRIPTION
**Pull request by roblabla**
Fixes #145 and #62

Here is what it looks like when it doesn't pass: 
Screenshot 2019-08-16 at 15 17 18

### CODE REVIEW

File: __tests__/validators/or.test.js [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    ]
+  }
+  let validation = await or.validate(createMockContext({title: 'Version 1'}), settings)
+  expect(validation.status).toBe('pass')
+})
+
+test('should return output of first task to pass when multiple are given', async() => {
+  const or = new Or()
+  const settings = {
+    do: 'or',
```
Comment (shine2lay) [inline __tests__/validators/or.test.js]: Is it possible to nest `do: 'or' ` ?
i.e 
``` 
do : or 
validate: 
  do : or
    validate: 
       do : A
       do : B
  do : C
```

Comment (roblabla) [reply __tests__/validators/or.test.js]: Should work, not sure why you’d do this though.

Comment (shine2lay) [reply __tests__/validators/or.test.js]: You are right, the scenario i proposed doesn't make sense, sorry about that. My thought were to make sure if and when we have 'and' validator, we can do nested 'and' and 'or' to create more complex validations. I should've taken a deeper look at your code first before asking.

**Comment by roblabla** [comment]
Added a test for the case mentioned by @shine2lay and some docs on the README. I think this is ready for review.

File: lib/validators/or.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            '',
+            validationSettings,
+            ERROR_MESSAGE
+          )
+        ],
+        validatorContext
+      )
+    }
+
+    const output = await Promise.all(promises)
```
[the remark this answers is not in the document]
Comment (roblabla) [reply lib/validators/or.js]: 🤔 the promises aren't supposed to throw an error. When the sub-promises are used directly, [processWorkflow](https://github.com/mergeability/mergeable/blob/a8eca8af88880118a8c5e69d8dc722b1df424a42/lib/flex.js#L64) also silently discards the error. I'm of the opinion that if an error is thrown, the Or validator should bubble it up, and processWorkflow should definitely be handling potential errors by having a `catch` clause there.

**Comment by roblabla** [comment]
There's already a screenshot of failing output in the issue description BTW ^^.

File: lib/validators/or.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (roblabla) [reply lib/validators/or.js]: I'll add a `.catch` to processWorkFlow tomorrow. This way all unexpected validator errors will bubble up and eventually get handled there. It's cleaner that way and more general.

**Comment by shine2lay** [comment]
@roblabla my mistake. I totally missed that.

File: lib/validators/or.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (roblabla) [reply lib/validators/or.js]: Hmm, so I've been prototyping with this for a bit, and uncaught errors are a bit hard to handle correctly in a generic way. I'll open a separate PR shortly with a prototype of how I think it should look like.

[the remark this answers is not in the document]
Comment (roblabla) [reply lib/validators/or.js]: #230

**Comment by shine2lay** [comment]
@roblabla I manually tested this, I found that the output is a bit hard to understand when you have multiple criteria. 
sample config
``` - do: or
        validate:
          - do: title
            must_exclude:
              regex: 'must_be_excluded_text'
            must_include:
              regex: 'must_be_included_text'
          - do: label
            or:
              - must_include:
                  regex: 'must_be_included_text'
              - begins_with:
                  regex: 'begins_with_text'
            begins_with:
              match: 'begins_with_text'
```
and i got this output
image

also, when one of the `or` condition passes, i get partial output which is a bit misleading I think
image

if you need more details, here is the repo where i tested this https://github.com/shine2lay/MergeableTest/pull/49

**Comment by roblabla** [comment]
Ah yeah, that's definitely not ideal. Didn't think of multiple criteria at all 😅 .

Partial output on success is "expected" although really confusing. Or validators return the first successful subvalidator, which is how I "trick" the higher-level function into passing the run.  Both problems are really symptoms of the same general issue though: Validators are given very little leeway to customize their output. If they want to show something as failed *without* failing the run, they're kinda stuck. So I think the next step will be to make the consolidated output structure a bit more flexible somehow (need to think about how to do that).

I won't have time to work on this for a couple weeks.

**Comment by shine2lay** [comment]
Since the output can be confusing, i add "Option #" in front of the description. Even this can get pretty gnarly when we have nested 'or/and' 
image

**Comment by jusx** [comment]
resolves #62

**Pull request merged**